quiche/src/dgram.rs

```diff
@@ -27,44 +27,46 @@
 use crate::Error;
 use crate::Result;
 
+use std::collections::BTreeMap;
 use std::collections::VecDeque;
 
 /// Keeps track of DATAGRAM frames.
 #[derive(Default)]
 pub struct DatagramQueue {
-    queue: Option<VecDeque<Vec<u8>>>,
+    queue: BTreeMap<u8, VecDeque<Vec<u8>>>,
     queue_max_len: usize,
     queue_bytes_size: usize,
 }
 
 impl DatagramQueue {
     pub fn new(queue_max_len: usize) -> Self {
         DatagramQueue {
-            queue: None,
+            queue: BTreeMap::new(),
             queue_bytes_size: 0,
             queue_max_len,
         }
     }
 
-    pub fn push(&mut self, data: Vec<u8>) -> Result<()> {
+    pub fn push(&mut self, data: Vec<u8>, urgency: u8) -> Result<()> {
         if self.is_full() {
             return Err(Error::Done);
         }
 
         self.queue_bytes_size += data.len();
         self.queue
-            .get_or_insert_with(Default::default)
+            .entry(urgency)
+            .or_insert_with(Default::default)
             .push_back(data);
 
         Ok(())
     }
 
     pub fn peek_front_len(&self) -> Option<usize> {
-        self.queue.as_ref().and_then(|q| q.front().map(|d| d.len()))
+        self.queue.iter().next().and_then(|(_, q)| q.front().map(|d| d.len()))
```

Comment:
Now that it's stable, https://doc.rust-lang.org/stable/std/collections/struct.BTreeMap.html#method.first_key_value is another option